File: Object-C/BlackFireIM/Example/ChatDetail/Controller/BFCallMessageCell.m
//
//  MSCallMessageCell.m
//  BlackFireIM
//
//  Created by benny wang on 2021/8/18.
//

#import "BFCallMessageCell.h"
#import "MSIMSDK-UIKit.h"


@implementation BFCallMessageCell

- (instancetype)initWithStyle:(UITableViewCellStyle)style reuseIdentifier:(NSString *)reuseIdentifier
{
    self = [super initWithStyle:style reuseIdentifier:reuseIdentifier];
    if (self) {
        
        _icon = [[UIImageView alloc]init];
        [self.bubbleView addSubview:_icon];
        
        _titleL = [[UILabel alloc] init];
        [self.bubbleView addSubview:_titleL];
    }
    return self;
}

- (BFCallMessageCellData *)callData
{
    return (BFCallMessageCellData *)self.messageData;
}

- (void)fillWithData:(BFCallMessageCellData *)data
{
    [super fillWithData:data];
    if (data.direction == MsgDirectionIncoming) {
        self.titleL.textColor = [UIColor d_colorWithColorLight:TText_Color dark:TText_Color];
    }else {
        self.titleL.textColor = [UIColor d_colorWithColorLight:[UIColor whiteColor] dark:[UIColor whiteColor]];
    }
    self.titleL.font = [UIFont systemFontOfSize:16];
    self.titleL.text = data.notice;
    self.icon.image = data.iconImage;
}

- (void)layoutSubviews
{
    [super layoutSubviews];
    self.titleL.frame = self.callData.noticeFrame;
    self.icon.frame = self.callData.iconFrame;
}

@end

File: Swift/MSIM-Swift/ChatDetail/C2C/BFCallMessageCell.swift
//
//  BFCallMessageCell.swift
//  MSIM-Swift
//
//  Created by benny wang on 2021/10/3.
//

import UIKit

class BFCallMessageCell: MSBubbleMessageCell {

    private var icon: UIImageView!
    
    private var titleL: UILabel!
    
    var callData: BFCallMessageCellData? {
        return self.messageData as? BFCallMessageCellData
    }
    
    override init(style: UITableViewCell.CellStyle, reuseIdentifier: String?) {
        super.init(style: style, reuseIdentifier: reuseIdentifier)
        
        icon = UIImageView()
        bubbleView.addSubview(icon)
        
        titleL = UILabel()
        bubbleView.addSubview(titleL)
    }
    
    override func fillWithData(data: MSMessageCellData) {
        super.fillWithData(data: data)
        if let callData = data as? BFCallMessageCellData {
            if data.direction == .inComing {
                titleL?.textColor = UIColor.d_color(light: MSMcros.TText_Color, dark: MSMcros.TText_Color)
            }else {
                titleL?.textColor = UIColor.d_color(light: .white, dark: .white)
            }
            titleL?.font = .systemFont(ofSize: 16)
            titleL?.text = callData.notice
            icon?.image = callData.iconImage
        }
    }
    
    override func layoutSubviews() {
        super.layoutSubviews()
        titleL?.frame = callData?.noticeFrame ?? .zero
        icon?.frame = callData?.iconFrame ?? .zero
    }
    required public init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
}
